File: January/Day008/solution.cpp
class Solution {
public:
    int maxDotProduct(vector<int>& nums1, vector<int>& nums2) {
        int m = nums1.size();
        int n = nums2.size();
        vector<int> prev(n, INT_MIN);
        vector<int> curr(n, INT_MIN);
        prev[0] = nums1[0] * nums2[0];
        for (int j = 1; j < n; j++) {
            prev[j] = max(prev[j-1], nums1[0] * nums2[j]);
        }
        for (int i = 1; i < m; i++) {
            curr[0] = max(prev[0], nums1[i] * nums2[0]);
            
            for (int j = 1; j < n; j++) {
                int curr_product = nums1[i] * nums2[j];
                int take_both = max(curr_product, prev[j-1] + curr_product);
                int skip_i = prev[j];
                int skip_j = curr[j-1];
                
                curr[j] = max({take_both, skip_i, skip_j});
            }
            
            swap(prev, curr);
        }
        
        return prev[n-1];
    }
};

File: January/Day008/solution.py
from typing import List

class Solution:
    def maxDotProduct(self, nums1: List[int], nums2: List[int]) -> int:
        m, n = len(nums1), len(nums2)
        # dp[i][j] represents max dot product using elements up to index i in nums1 
        # and index j in nums2
        dp = [[float('-inf')] * n for _ in range(m)]
        dp[0][0] = nums1[0] * nums2[0]
        for j in range(1, n):
            dp[0][j] = max(dp[0][j-1], nums1[0] * nums2[j])
        for i in range(1, m):
            dp[i][0] = max(dp[i-1][0], nums1[i] * nums2[0])
        for i in range(1, m):
            for j in range(1, n):
                curr_product = nums1[i] * nums2[j]
                take_both = max(curr_product, dp[i-1][j-1] + curr_product)
                skip_i = dp[i-1][j]
                skip_j =  dp[i][j-1]
                
                dp[i][j] = max(take_both, skip_i, skip_j)
        
        return dp[m-1][n-1]